retry/src/main/java/com/iluwatar/retry/RetryWithExponentialBackoff.java

```diff
@@ -0,0 +1,117 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.iluwatar.retry;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Predicate;
+import java.lang.Math;
+import java.util.Random;
+
+/**
+ * Decorates {@link BusinessOperation business operation} with "retry" capabilities.
+ *
+ * @author George Aristy (george.aristy@gmail.com)
+ * @param <T> the remote op's return type
+ */
+
+public final class RetryWithExponentialBackoff<T> implements BusinessOperation<T> {
+    private final BusinessOperation<T> op;
+    private final int maxAttempts;
+    private final long delay;
+    private final long maxDelay;
+    private final AtomicInteger attempts;
+    private final Predicate<Exception> test;
+    private final List<Exception> errors;
+
+    /**
+     * Ctor.
+     *
+     * @param op the {@link BusinessOperation} to retry
+     * @param maxAttempts number of times to retry
+     * @param delay delay (in milliseconds) between attempts
+     * @param ignoreTests tests to check whether the remote exception can be ignored. No exceptions
+     *     will be ignored if no tests are given
+     */
+    @SafeVarargs
+    public RetryWithExponentialBackoff(
+            BusinessOperation<T> op,
+            int maxAttempts,
+            long maxDelay,
+            Predicate<Exception>... ignoreTests
+    ) {
+        this.op = op;
+        this.maxAttempts = maxAttempts;
+        this.maxDelay = maxDelay;
+        this.attempts = new AtomicInteger();
+        this.test = Arrays.stream(ignoreTests).reduce(Predicate::or).orElse(e -> false);
+        this.errors = new ArrayList<>();
+    }
+
+    /**
+     * The errors encountered while retrying, in the encounter order.
+     *
+     * @return the errors encountered while retrying
+     */
+    public List<Exception> errors() {
+        return Collections.unmodifiableList(this.errors);
+    }
+
+    /**
+     * The number of retries performed.
+     *
+     * @return the number of retries performed
+     */
+    public int attempts() {
+        return this.attempts.intValue();
+    }
+
+    @Override
+    public T perform() throws BusinessException {
+        do {
+            try {
+                return this.op.perform();
+            } catch (BusinessException e) {
+                this.errors.add(e);
+
+                if (this.attempts.incrementAndGet() >= this.maxAttempts || !this.test.test(e)) {
+                    throw e;
+                }
+
+                try {
+                    long testDelay = Math.pow(2, this.attempts)*1000 + Math.round(Math.random()*1000);

```

Comment:
I don't think using `Math.random()` will give increasing backoff. We can think of using fibonacci series for delay.
Something similar to https://carlosbecker.com/posts/exponential-backoff-java8/

Reply:
Hi, thanks for responding! :D
Actually, I was trying to do it the way it was suggested in the link, they basically added a random number to 2^number_of_attempts. I think Fibonacci should work too though. Should it be changed to that?

Reply:
@AnaghaSasikumar adding random number to 2^no_of_attempts doesn't guarantee that back-off is increasing in order.

Reply:
That part will be < 1s, Math.random() generates a random double between 0 and 1.